File: limo/limo_learnning/src/limo_action_client.cpp
//导入头文件
#include <iostream>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include <geometry_msgs/msg/twist.hpp>
#include <std_msgs/msg/bool.hpp>
#include <limo_msgs/action/limo_action.hpp>

using namespace std;
using namespace std::placeholders;
// 定义一个LimoActionClient类，继承自rclcpp::Node
class LimoActionClient : public rclcpp::Node
{

  public:
    // 使用别名简化limo_msgs::action::LimoAction
    using Limoaction = limo_msgs::action::LimoAction;
    // 使用别名简化rclcpp_action::ClientGoalHandle<Limoaction>
    using ClientGoal = rclcpp_action::ClientGoalHandle<Limoaction>;
    // 构造函数，初始化节点名称为"limo_action"
    explicit LimoActionClient(const rclcpp::NodeOptions & node_options = rclcpp::NodeOptions())
    : Node("limo_action",node_options)
    {
      // 创建动作客户端
      this->action_client = rclcpp_action::create_client<Limoaction>(
        this->get_node_base_interface(),
        this->get_node_graph_interface(),
        this->get_node_logging_interface(),
        this->get_node_waitables_interface()
        , "limo_action");
    }
    //发送目标函数
    void send_goal()
    {
    if (!action_client->wait_for_action_server(std::chrono::seconds(20))) {
        RCLCPP_ERROR(this->get_logger(), "Action server not available after waiting");
        return ;
      }

    auto send_goal_opt = rclcpp_action::Client<Limoaction>::SendGoalOptions();
    send_goal_opt.goal_response_callback = 
      std::bind(&LimoActionClient::response_callback,this,_1); // 绑定目标响应回调函数
    send_goal_opt.feedback_callback = 
      std::bind(&LimoActionClient::feedback_callback,this,_1,_2); // 绑定反馈回调函数
    send_goal_opt.result_callback = 
      std::bind(&LimoActionClient::result_callback,this,_1);  // 绑定结果回调函数
    // 创建目标信息，并填上数据
    auto goal_msg = Limoaction::Goal();
    goal_msg.x = 1;
    goal_msg.y = 0;
    goal_msg.z = 0.5;


    RCLCPP_INFO(this->get_logger(), "Sending goal");
    // 发送目标
    this->action_client->async_send_goal(goal_msg,send_goal_opt);
    }

  private:
    rclcpp_action::Client<Limoaction>::SharedPtr action_client;
  // 目标响应函数
  void response_callback (ClientGoal::SharedPtr res_msg)
  {
    if (!res_msg)
    {
        RCLCPP_ERROR(this->get_logger(), "Client: Goal was rejected by server");
        rclcpp::shutdown(); // Shut down client node
    }
    else
    {
        RCLCPP_INFO(this->get_logger(), "Client: Goal accepted by server, waiting for result");
    }

  }

    // 创建处理周期反馈消息的回调函数
    void feedback_callback(
        ClientGoal::SharedPtr,
        const std::shared_ptr<const Limoaction::Feedback> feedback_message)
      {
          // 将反馈信息写入字符串流
          std::stringstream ss;
          ss << "Client: Received feedback: "<< feedback_message->status;
          RCLCPP_INFO(this->get_logger(), "%s", ss.str().c_str());
      }

      // 创建一个收到最终结果的回调函数
    void result_callback(const ClientGoal::WrappedResult & result_msg)
    {
        switch (result_msg.code)
        {
            case rclcpp_action::ResultCode::SUCCEEDED:
                break;
            case rclcpp_action::ResultCode::ABORTED:
                RCLCPP_ERROR(this->get_logger(), "Client: Goal was aborted"); // 输出目标被中止信息
                rclcpp::shutdown(); // 关闭客户端节点
                return;
            case rclcpp_action::ResultCode::CANCELED:
                RCLCPP_ERROR(this->get_logger(), "Client: Goal was canceled");  // 输出目标被取消信息
                rclcpp::shutdown(); 
                return;
            default:
                RCLCPP_ERROR(this->get_logger(), "Client: Unknown result code");  // 输出未知结果码信息
                rclcpp::shutdown();
                return;
        }
            // 输出结果信息
            RCLCPP_INFO(this->get_logger(), "Client: Result received: %s",result_msg.result->success ? "false" : "true");
        rclcpp::shutdown();         
    }

}
;

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  // 创建limo_action_client节点
  auto action_client = std::make_shared<LimoActionClient>(); 
  // 发送目标
  action_client->send_goal();

  rclcpp::spin(action_client);

  rclcpp::shutdown();

  return 0;
}

File: limo/limo_learnning_py/limo_learnning_py/limo_action_client.py
import rclpy  # 导入ROS 2 Python客户端库
from rclpy.node import Node  # 导入节点类
from rclpy.action import ActionClient  # 导入动作客户端类
# from rclpy.action import ClientGoalHandle  # 导入客户端目标句柄类
from limo_msgs.action import LimoAction  # 导入自定义动作消息类型
import sys  # 导入系统模块

class LimoActionClient(Node):  # 定义LimoActionClient类，继承自Node

    def __init__(self):  # 构造函数
        super().__init__('limo_action_client')  # 初始化节点名称为'limo_action_client'
        self.action_client = ActionClient(self, LimoAction, 'limo_action')  # 创建动作客户端

    def send_goal(self):  # 定义发送目标的函数
        if not self.action_client.wait_for_server(timeout_sec=20.0):  # 等待动作服务器20秒
            self.get_logger().error('Action server not available after waiting')  # 如果服务器不可用，记录错误日志
            rclpy.shutdown()  # 关闭节点
            return  # 返回退出函数

        goal_msg = LimoAction.Goal()  # 创建目标消息
        goal_msg.x = 1.0  # 设置目标x坐标
        goal_msg.y = 0.0  # 设置目标y坐标
        goal_msg.z = 0.5  # 设置目标z坐标

        self.get_logger().info('Sending goal')  # 记录发送目标的日志

        self.action_client.wait_for_server()  # 再次等待动作服务器
        # 异步发送目标，并绑定反馈回调函数
        # self.action_client.send_goal_async(goal_msg, feedback_callback=
        # self.feedback_callback).add_done_callback(self.goal_response_callback)

        self.send_goal_future = self.action_client.send_goal_async(goal_msg,
            feedback_callback=self.feedback_callback)
        self.send_goal_future.add_done_callback(self.goal_response_callback)

    def goal_response_callback(self, future):  # 定义目标响应回调函数
        goal_handle = future.result()  # 接受结果
        if not goal_handle.accepted:  # 如果目标未被接受
            self.get_logger().error('Client: Goal was rejected by server')  # 记录目标被拒绝的错误日志
            rclpy.shutdown()  # 关闭节点
            return  # 返回退出函数

        self.get_logger().info('Client: Goal accepted by server, waiting for result')  # 记录目标被接受的日志
        # goal_handle.get_result_async().add_done_callback(self.result_callback)  # 异步获取结果，并绑定结果回调函数
        self.get_result_future = goal_handle.get_result_async()
        self.get_result_future.add_done_callback(self.result_callback)

    def feedback_callback(self, feedback_message):  # 定义反馈回调函数
        feedback = feedback_message.feedback
        self.get_logger().info(f'Client: Received feedback: {feedback.status}')  # 记录收到的反馈信息

    def result_callback(self, future):  # 定义结果回调函数
        result = future.result().result  # 获取结果
        self.get_logger().info('Result: %d' % result.success)

def main(args=None):  # 定义主函数
    rclpy.init(args=args)  # 初始化rclpy
    action_client = LimoActionClient()  # 创建LimoActionClient对象
    action_client.send_goal()  # 发送目标
    rclpy.spin(action_client)  # 保持节点运行

if __name__ == '__main__':  # 如果是主程序入口
    main()  # 调用主函数
